docs/rfc/rfc-020-block-structure-consolidation.md

```diff
@@ -0,0 +1,363 @@
+
+#### ProofOfLockRound
+
+The *proof of lock round* is the round of consensus for a height in which the
+Tendermint algorithm observed a super majority of voting power on the network for
+a block.
+
+Including this value in the block will allow validation of currently
+un-validated metadata. Specifically, including this value will allow Tendermint
+to validate that the `ProposerAddress` in the block is correct. Without knowing
+the locked round number, Tendermint cannot calculate which validator was supposed
+to propose a height. Because of this, our [validation logic][proposer-check] does not check that
+the `ProposerAddress` included in the block corresponds to the validator that
+proposed the height. Instead, the validation logic simply checks that the value
+is an address of one of the known validators.
+
+Currently, we include this value in the `Commit` for height `H` which is
+written into the block at height `H+1`. Tendermint can add this value directly
+into the block instead of delaying it until the next block.
+
+### Additional possible updates
+
+#### Updates to storage
+
+Currently we store the [every piece of each block][save-block] in the `BlockStore`.
+I suspect that this has lead to some mistakes in reasoning around the merits of
+consolidating fields in the block. We could update the storage scheme we use to
+store only some pieces of each block and still achieve a space savings without having
+to change the block structure at all.
+
+The main way to achieve this would be by _no longer saving data that does not change_.
+At each height we save a set of data that is unlikely to have changed from the
+previous height in the block structure, this includes the `ValidatorAddress`es,
+the `ValidatorsHash`, the `ChainID`. These do not need to be saved along with
+_each_ block. We could easily save the value and the height at which the value
+was updated and construct each block using the data that existed at the time.
+
+This document does not make any specific recommendations around storage since
+that is likely to change with upcoming improvements to to the database infrastructure.
+However, it's important to note that removing fields from the block for the
+purposes of 'saving space' may not be that meaningful. We should instead focus
+our attention of removing fields from the block that are no longer needed
+for correct functioning of the protocol.
+
+#### Updates to propagation
+
+Block propagation suffers from the same issue that plagues block storage, we
+propagate all of the contents of each block proto _even when these contents are redundant
+or unchanged from previous blocks_. For example, we propagate the `ValidatorAddress`es
+for each block in the `CommitSig` structure even when it never changed from a
+previous height. We could achieve a speed-up in many cases by communicating the
+hashes _first_ and letting peers request additional information when they do not
+recognize the communicated hash. 
```

Comment:
Yeah, I think it would make the architecture much cleaner if we had a content addressing layer. To clarify though, I have no concrete plans to actually do this in the next release or two. I think it's something we can / should consider post 1.0. I think there are more low hanging fruit in the gossip layer to hit before implementing something like content addressing into Tendermint.